Automattic/sensei
# Implement the quiz pagination frontend

Fixes #4480 

### Changes proposed in this Pull Request

This PR adds the quiz pagination frontend UI and refactors some of the default quiz action buttons code. It does not implement the session needed for the quiz form submission to work properly, so no need to test that.

### Testing instructions

* Enable the quiz pagination feature:
```php
add_filter( 'sensei_feature_flag_quiz_pagination', '__return_true' );
```
* Make a new course with a lesson.
* Make a quiz with multiple questions (>=8).
* Enable the pagination by updating the quiz pagination settings (see #4387).
* Enroll in the course.
* Open the quiz in the frontend.
* Don't use the save/reset/complete buttons. They will be fixed when the quiz session is implemented (#4392).
* Make sure the question listing is paginated.
* Make sure the numeric pagination and the next/prev buttons work.
* Check if the question numbers are correct.
* Edit the quiz and enable the `Allow Retakes` setting and make sure the `Reset` button is present in the frontend.
* Make sure you see the `Complete` button on the last page.
* Make sure you have at least 8 pages and go to page 4 to see the full width of the numeric pagination.
* Make sure the pagination looks good on the more popular WP themes.
* Make sure the pagination looks good on mobile.
* Make sure the question listing is paginated even if not enrolled in the course.
* Make the quiz pagination per page number bigger than the questions count and make sure the pagination UI is not loading. You should see the default action buttons instead.
* Disable the quiz pagination and make sure the default quiz action buttons work as expected.

### New Hooks

* `sensei_quiz_pagination_args` - Filters the quiz questions paginate links args.
* `sensei_the_question_number` - Filters the number of the current question within the quiz question loop.

### Changed Templates

* `single-quiz.php` - Explicitly provide the question number.

### Screenshots

Desktop:
<img width="662" alt="Screenshot on 2021-12-14 at 15-51-34" src="https://user-images.githubusercontent.com/1612178/146012162-f395e4b4-6905-459d-b75a-15996d744036.png">
<img width="851" alt="Screenshot on 2021-12-14 at 15-54-04" src="https://user-images.githubusercontent.com/1612178/146012191-d13eb774-096f-4bf8-8f11-51869ce8487a.png">


Mobile:
<img width="353" alt="Screenshot on 2021-12-14 at 15-55-39" src="https://user-images.githubusercontent.com/1612178/146012217-562b23d1-bfc7-4c00-a019-4aab2799706b.png">
<img width="341" alt="Screenshot on 2021-12-14 at 15-54-47" src="https://user-images.githubusercontent.com/1612178/146012235-c9cf39d4-2402-43ef-a2b5-2f4e5cecc42a.png">
